engine/src/test/java/io/camunda/zeebe/engine/processing/scheduled/DueDateCheckerTest.java

```diff
@@ -0,0 +1,126 @@
+
+  @Test
+  void shouldNotScheduleTwoTasks() {
+    // given
+    final var dueDateChecker =
+        new DueDateChecker(100, (builder) -> new ProcessTriggerResult(0L, () -> {}));
+    final var mockScheduleService = mock(DueDateScheduler.class);
+
+    dueDateChecker.onRecovered(mockScheduleService);
+    verify(mockScheduleService, timeout(1000)).runDelayed(eq(Duration.ZERO), any(Task.class));
+    Mockito.clearInvocations(mockScheduleService);
+
+    // when
+    final var currentTimeMillis = System.currentTimeMillis();
+    dueDateChecker.schedule(currentTimeMillis + 1000); // in one second
+    dueDateChecker.schedule(currentTimeMillis + 1000); // in one second
+
+    // then
+    verify(mockScheduleService, timeout(1000)).runDelayed(any(), any(Task.class));
+  }
+
+  @Test
+  void shouldScheduleForAnEarlierTasks() {
+    // given
+    final var dueDateChecker =
+        new DueDateChecker(100, (builder) -> new ProcessTriggerResult(0L, () -> {}));
+    final var mockScheduleService = mock(DueDateScheduler.class);
+
+    dueDateChecker.onRecovered(mockScheduleService);
+    verify(mockScheduleService, timeout(1000)).runDelayed(eq(Duration.ZERO), any(Task.class));
+    Mockito.clearInvocations(mockScheduleService);
+
+    // when
+    final var currentTimeMillis = System.currentTimeMillis();
+    dueDateChecker.schedule(currentTimeMillis + 1000); // in one second
+    dueDateChecker.schedule(currentTimeMillis + 100); // in 100 millis
+
+    // then
+    verify(mockScheduleService, timeout(1000).times(2)).runDelayed(any(), any(Task.class));
+  }
+
+  @Test
+  void shouldScheduleNextTask() {
+    // given
+    final var now = Instant.now();
+    final var nextDueDate = now.plusMillis(5000);
+    final var scheduler = new TestDueDateScheduler();
+    final var dueDateChecker =
+        new DueDateChecker(
+            100,
+            (builder) -> new ProcessTriggerResult(nextDueDate.toEpochMilli(), () -> {}),
+            now::toEpochMilli);
+
+    // when
+    dueDateChecker.onRecovered(scheduler);
+    scheduler.scheduledTasks.remove(0).task.execute(mock(TaskResultBuilder.class));
+
+    // then
+    assertThat(scheduler.scheduledTasks)
+        .map(ScheduledTask::delay)
+        .containsExactly(Duration.ofSeconds(5));
+  }
+
+  @Test
+  void shouldPersistCache() {
+    // given
+    final var now = Instant.now();
+    final var nextDueDate = now.plusMillis(5000);
+    final var scheduler = new TestDueDateScheduler();
+    final var cachePersisted = new MutableBoolean();
+    final var dueDateChecker =
+        new DueDateChecker(
+            100,
+            (builder) ->
+                new ProcessTriggerResult(
+                    nextDueDate.toEpochMilli(), () -> cachePersisted.set(true)),
+            now::toEpochMilli);
+
+    // when
+    dueDateChecker.onRecovered(scheduler);
+    scheduler.scheduledTasks.remove(0).task.execute(mock(TaskResultBuilder.class));
+
+    // then
+    assertThat(cachePersisted.get()).isTrue();
```

Comment:
Not sure whether this is optimal since we couple this here heavily with the implementing. The behavior of not scheduling the task twice should be enough.

Reply:
Probably. I added it because initially I wanted to test it's persisted at the very end after the task result is built, but we can probably remove it.